src/smallbox.rs

```diff
@@ -177,49 +164,39 @@ impl<T: ?Sized, Space> SmallBox<T, Space> {
     /// ```
     #[inline]
     pub fn is_heap(&self) -> bool {
-        self.ptr.as_ptr().cast::<u8>() != INLINE_SENTINEL
+        !self.ptr.is_null()
     }
 
     unsafe fn new_copy<U>(val: &U, metadata_ptr: *const T) -> SmallBox<T, Space>
     where U: ?Sized {
-        let layout = Layout::for_value::<U>(val);
-        let space_layout = Layout::new::<Space>();
+        let size = mem::size_of_val(val);
+        let align = mem::align_of_val(val);
 
         let mut space = MaybeUninit::<UnsafeCell<Space>>::uninit();
 
-        let (ptr_this, val_dst): (*mut u8, *mut u8) =
-            if layout.size() <= space_layout.size() && layout.align() <= space_layout.align() {
-                // Stack.
-                (INLINE_SENTINEL, space.as_mut_ptr().cast())
-            } else if layout.size() == 0 {
-                // ZST with alignment greater than Space, which will behave like being stored on
-                // heap but will not actually allocate.
-                (
-                    sptr::without_provenance_mut(layout.align()),
-                    sptr::without_provenance_mut(layout.align()),
-                )
-            } else {
-                // Heap.
-                let layout = layout
-                    // Safety: MIN_ALIGNMENT is 2, which is a valid power-of-two alignment.
-                    .align_to(MIN_ALIGNMENT)
-                    .unwrap_or_else(|_| unreachable_unchecked());
-                let heap_ptr = alloc::alloc(layout);
-
-                if heap_ptr.is_null() {
-                    handle_alloc_error(layout)
-                }
+        let (ptr_this, val_dst): (*mut u8, *mut u8) = if size == 0 {
+            (
+                sptr::without_provenance_mut(align),
+                sptr::without_provenance_mut(align),
+            )
+        } else if size > mem::size_of::<Space>() || align > mem::align_of::<Space>() {
```

Comment:
With this approach ZST would be properly stored on stack.

```suggestion
        let (ptr_this, val_dst): (*mut u8, *mut u8) =
            if layout.size() <= space_layout.size() && layout.align() <= space_layout.align() {
                // Layout requirement satisfied; store on stack.
                (ptr::null_mut(), space.as_mut_ptr().cast())
            } else if layout.size() == 0 {
                // ZST with unsatisfied alignment; pretend to store on the heap.
                (
                    sptr::without_provenance_mut(layout.align()),
                    sptr::without_provenance_mut(layout.align()),
                )
            } else {
                // Otherwise, allocate on the heap.
                let layout = Layout::for_value::<U>(val);
                let heap_ptr = alloc::alloc(layout);

                if heap_ptr.is_null() {
                    handle_alloc_error(layout)
                }

                (heap_ptr, heap_ptr)
            };
```

Reply:
We could, but since ZSTs on the "heap" don't allocate in the first place, the only reason I can see to do this would be to affect the result of `is_heap`. Currently ZSTs will be consistently on the "heap". With your suggestion, non-over aligned ZSTs will be on the "stack", and sufficiently over aligned ZSTs will end up on the "heap".

How does this affect the user though? I don't understand where a user would use `is_heap`, so I don't know what behavior it should have. Do you know how `is_heap` is used?

Reply:
Thinking about this a bit more, your suggestion does cause ZSTs to behave equivalently to non-ZSTs. Whatever `is_heap` is used for, I suspect that consistency between ZSTs and non-ZSTs is good.

Note that as of #47, and thus the last two stable versions, the behavior of `is_heap` is to always store ZSTs on the "heap". Before #47, ZSTs were always stored on the "stack". So this suggestion does change the behavior of `is_heap`.

I do think that consistency is a good reason to make the change though, so I'll accept the suggestion tomorrow if I don't hear back from you.

Reply:
Yes, ZST is tricky since it's neither stored on heap nor stack. Any of these would make the rule of whether the smallbox optimzation had taken effect complex. And with this change, the rule will become simple and consistent:  store on stack when both size and alignment fit.